### PR TITLE
135 - IdsPopup Layout Logic and Improvements

### LINKED ISSUE #135: IdsPopup: Complete layout logic
**Describe the bug**

Complete remaining work on ids-popup:

- [x] Bring the Popup component into alignment with recent changes (use `super`, use the same `state` object, etc)
- [x] Support the nudge/flip logic that was present in 4.x (can be tested on toolbar or theme switcher)
- [x] Allow for placement in IdsContainer (and other elements)
- [x] In addition there is a potential problem on https://github.com/infor-design/enterprise-wc/blob/main/src/ids-popup-menu/ids-popup-menu.js#L236 - if you have a Popupmenu in the page and something that requires a click then the click event is prevented default on and the click wont fire. I put in a fix for this but i think the target should be targeted for click here but this may require further revision
- [x] Another issue is that this line https://github.com/infor-design/enterprise-wc/blob/main/src/ids-popup/ids-popup.js#L637 will run if you intend to have an alignTarget initially when the component inits this calls a `getBoundingClientRect` and that does a reflow. Should investigate if this should be called. I wrapped it in a `if (this.visible)` which seems correct in the tooltip PR but this can use more investigation

**Additional context**
- 100% test coverage
- DOCS, CHANGE log ect. 
- https://github.com/infor-design/enterprise-wc/blob/master/doc/CHECKLIST.MD
- themes 

See completed components like card, tag, badge, hyperlink as reference

### LINKED ISSUE #198: Tooltip: Animation jiggle when showing and hiding
**Describe the bug**
Seeing a newly noticed animation problem on tooltip. Noticed it did not happen on my branch so was one of the PR's in the this sprint.

**To Reproduce**
1. Go to http://localhost:4300/ids-tooltip
2. hover button -> see it jiggle a little
3. hover out of button -> see it jank off the page

**Expected behavior**
As it was before

**Additional context**
Noticed it did not happen on my branch so was one of the PR's in the last 10 days. Also reverted ids-popup and thats not it. But seems something in the context of the setting of properties.

Can be fixed by changing:

```
    this.popup.alignTarget = null;
    this.removeAttribute('visible');
    this.#hide();
 ```

to

```
    this.#hide();
    this.popup.alignTarget = null;
    this.removeAttribute('visible');
```

But want to make sure there isnt something worse happening.

### PR DESCRIPTION
**Pull request by EdwardCoyle**
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR makes some general improvements to the IdsPopup component:
- Fixes a bug reported where the arrow applied to some popups would visually "jump" to the correct position after the Popup was rendered.
- Refactors the lifecycle overall to not call `refresh()` as often as before, improving speed and cutting down on DOM repaints.
- Removes `safeSetAttribute/safeRemoveAttribute` in favor of better checks on setter values to prevent double-setting of properties from occurring.
- Adds a `containingElem` setting that can be configured to keep a Popup confined within the boundaries of another element on the page.  By default, this is the IdsContainer present on all pages, so all Popups will remain within those boundaries.
- Adds a `bleed` setting, which can be used to disable/enable the containment behavior.
- Ports over the `nudge` and `flip` behaviors from the 4.x Place component, and directly integrates them into the `placeAtCoords` and `placeAgainstParent` routines.

**Related github/jira issue (required)**:
Closes #135 
Closes #198 

**Steps necessary to review your pull request (required)**:

_To test #198:_
- Open http://localhost:4300/ids-menu-button 
- Open both menus and watch the arrow.  It should open and remain in the correct placement, beneath each component.

_To test containment:_
- Open http://localhost:4300/ids-popup-menu
- Right-click (or CTRL+Click) with the mouse cursor placed closely to the edge of the page  An IdsPopupMenu should open, and should not be cut off by the browser viewport.
- Open the submenus, and the same should be true for each.

_Another containment test:_
- Open http://localhost:4300/ids-popup/test-sandbox
- Under the "Alignment Targets" fieldset, click "None (Use Coordinates)".  The Popup will move to the top/left corner of the browser viewport.
- Under the "Containment" fieldset, click "Test Container".  The Popup will move to the top/left of the gray test container element in the page.
- Under the "Coordinates" fieldset, set the X/Y coordinates each to at least 600px.  The Popup will move to the bottom/right corner of the gray test container, but it will not cross the edges.
- Under the "Other" fieldset, click "Allow Bleeds".  The Popup will reposition and cross the edges of the gray test container.

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

File: app/ids-menu-button/test-menus-on-edge.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import IdsMenuButton from '../../src/ids-menu-button/ids-menu-button';
+import IdsPopupMenu from '../../src/ids-popup-menu/ids-popup-menu';
+import IdsMenuGroup from '../../src/ids-menu/ids-menu-group';
+import IdsMenuHeader from '../../src/ids-menu/ids-menu-header';
+import IdsMenuItem from '../../src/ids-menu/ids-menu-item';
+import IdsSeparator from '../../src/ids-menu/ids-separator';
+import IdsPopup from '../../src/ids-popup/ids-popup';
+import IdsToolbar from '../../src/ids-toolbar/ids-toolbar';
+
+document.addEventListener('DOMContentLoaded', () => {
```
Comment (tmcconechy) [inline app/ids-menu-button/test-menus-on-edge.js:10]: unneeded?

**Review by tmcconechy — COMMENTED** [review]
1 . See some icon problems now on http://localhost:4300/ids-menu-button
Screen Shot 2021-06-17 at 9 47 00 AM
2. on http://localhost:4300/ids-popup-menu if i right click on the right side its not working as i expected? And the icons in the menu are off

Screen Shot 2021-06-17 at 9 50 01 AM
Screen Shot 2021-06-17 at 9 49 51 AM
3 on http://localhost:4300/ids-popup/test-sandbox i see extra scroll bars and no bleed button

This almost seems like im missing something? But i checked twice.

**Comment by EdwardCoyle** [comment]
@tmcconechy not sure what happened but I'm not seeing those situations when I run this on local:

| Screen Shot 2021-06-17 at 11 33 59 AM  Screen Shot 2021-06-17 at 11 35 36 AM

Wondering if you can recheck sometime soon?  I did remove the unneeded test.

About the Percy tests:  I think the pages are all rendering wider because of the newly-displayed Theme Switcher button.  Some of the examples need adjusting so the grid layout doesn't stretch them too far, so I'll push a fix for those soon, but I think that Percy run could probably be accepted afterward.

**Comment by tmcconechy** [comment]
@EdwardCoyle - yes all that was my bad... It is working very well. Just some minor nits (really bugs actually)

1. Sorry about the merge conflicts i merged some other branches
2. In dark mode the arrow is the wrong color

Screen Shot 2021-06-18 at 3 36 32 PM

3. For the theme switcher if you pick a theme it doesnt close. It seems like it doesnt close selcting submenus but it does close selecting the main menu (try switching theme)
4. The submenu arrows are a bit out of alignment
Screen Shot 2021-06-18 at 3 36 58 PM


If you want to punt on those i can make an issue of i can look into them next week while you are away. They see like simple things

**Comment by EdwardCoyle** [comment]
I'm not sure about #4, not sure what would've changed here to target the icons.  

I was noticing some issues with the Popup Menu usability in different situations as well.  There's some combinations of the button/menu being in light/shadow DOM that have strange results for some of the events.  I think it's worth punting those to another issue and looking at separately.  I can do it when I return if you like

**Review by tmcconechy — APPROVED** [review]

**Review by deep7102 — APPROVED** [review]

**Pull request merged**